### ISSUE TITLE
design(kv): prove bit-exact middle-span Evict survives paged/block KV (Evict-on-paged tension)

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#33]
_Part of #50 — large-scale disaggregated serving (dual-track RIDE + NATIVE)._

## Context

This is a **Track B (NATIVE)** design issue. fak's own disaggregated engine needs a paged/block KV allocator to reach SOTA KV occupancy (sibling: *feat(kv): paged/block KV allocator carrying the bit-exact Evict value-add*). But paging breaks the one assumption fak's headline KV value-add is built on — and we must prove the value-add survives **before** we build the allocator, or paged KV is just a worse-resourced re-implementation of PagedAttention with none of fak's differentiation.

**The value-add today (verified):** `internal/model/kv.go:60-103` `Evict(from, n)` removes a contiguous span from EVERY layer and compacts survivors so the cache is byte-for-byte what it would be if the span had never been seen. The correctness hinges on two contiguity-dependent facts in the current code:
- KV is stored as one **flat per-layer slice** — `K[l]`, `Kraw[l]`, `V[l]` are `[]float32` indexed by `position*kvStride` (kv.go:18-26, :74-80). Eviction is `append(K[l][:from*w], K[l][end*w:]...)` — a contiguous compaction.
- Survivors that sat *after* the span are **re-RoPE'd in a single rotation** from the pre-RoPE `Kraw[i]` at their NEW index `i` (kv.go:85-101). The comment (kv.go:28-31) is explicit that a single rotation at the new position is bit-exact, whereas composing two rotations drifts ~1e-6 and can flip a greedy token. The GLM-DSA path mirrors this (`evictGLMDsa`, kv.go:105-117).

**Why paging breaks it:** a block-paged allocator (fixed-size blocks + per-sequence block table, vLLM V1 `BlockManager`/`KVCacheManager`; SGLang token-attention paged pool over RadixAttention) stores positions in **non-contiguous physical blocks**. "Remove span [from,from+n) and re-index every survivor to a new contiguous position" is no longer a slice splice — survivors keep their physical block slots, and re-indexing every survivor to a new logical position means re-RoPE-ing K across block boundaries, possibly for blocks shared copy-on-write with other sequences. The `== never-saw-it` guarantee and the single-rotation re-RoPE invariant must be re-derived for a layout where physical position != logical position.

**SOTA framing (honest):** this is **not a parity item** — vLLM `reset_prefix_cache` and SGLang `flush_cache` only whole-prefix-reset; neither has an exact-span eviction API (`enginecache.go` hardcodes `SupportsExactSpan=false` for both). So no SOTA design exists to copy. This issue exists precisely because fak's exact-span Evict is the reason a NATIVE kernel-owned paged KV can be strictly **better** than a ridden engine — and the Track A counterpart (riding vLLM/SGLang) structurally **cannot** offer it (it degrades to whole-prefix flush). If this design fails, the honest answer is that paged KV should ride external PagedAttention and fak keeps exact-span only on the contiguous single-node path.

## Scope

Resolve the Evict-on-paged tension **on paper + in a prototype**, before the paged allocator lands:

- Specify exact-span eviction over a block-paged layout that preserves the `== never-saw-it` guarantee: how a logical span maps to (possibly partial) physical blocks, how survivors are re-indexed, and what happens at block boundaries.
- Specify the **survivor re-RoPE strategy across block boundaries**: when physical position decouples from logical position, define where/when the single-rotation-from-`Kraw` happens so it stays bit-exact (no two-rotation composition). Decide whether `Kraw` is stored per-block, and the COW-sharing interaction (a re-RoPE must not mutate a block shared with a sequence that did NOT evict).
- Build a prototype (can be a model-level harness, not the full allocator) that performs a **mid-span** evict on a paged block layout and asserts the resulting logits/cache are **bit-identical** to the existing contiguous `Evict`.
- State the **FP8/INT8-KV interaction** explicitly: the current bit-exact re-RoPE depends on f32 `Kraw` (kv.go:21). A quantized-KV path (sibling parity row, currently ABSENT) cannot drop in without either preserving an f32 `Kraw` shadow or re-justifying the precision. This issue must either carry that constraint or **defer** quantized-KV with a written reason.

## Acceptance

- [ ] Written design for exact-span eviction over block-paged KV that preserves the `== never-saw-it` guarantee (committed under `docs/`)
- [ ] Survivor re-RoPE-across-block-boundaries strategy specified, including the COW-shared-block case and whether `Kraw` is stored per-block
- [ ] Prototype demonstrates a mid-span evict on paged blocks is **bit-identical** to the contiguous `kv.go:60` `Evict` (test/harness committed, run via `.\fak\test.ps1`)
- [ ] GLM-DSA path (`evictGLMDsa`, kv.go:105) covered or explicitly scoped out with reason
- [ ] FP8/INT8-KV interaction explicitly stated — re-justifies preserving f32 `Kraw`, or defers quantized KV with a written reason
- [ ] Explicit go/no-go recommendation: if exact-span cannot survive paging affordably, say so and route paged KV to ride external PagedAttention (keeping exact-span on the contiguous single-node path)

## Dependencies

- Parent epic: *epic(serving): large-scale disaggregated serving — dual-track (RIDE + NATIVE) over one shared spine*
- Hard prerequisite-of (this design must land first): *feat(kv): paged/block KV allocator carrying the bit-exact Evict value-add* and *feat(governance): wire L3 KV-governance referee + native exact-span eviction (build-on #504/#532/#129/#135)*
- Track A counterpart this contrasts with: *feat(serving): orchestrate external P/D disaggregation + govern KV-transport bridge (NIXL/Mooncake/LMCache)* — where exact-span degrades to whole-prefix flush (`enginecache.go` `SupportsExactSpan=false`)
- Related existing issues: **#504**/**#532** (L3 KV-governance referee / native exact-span eviction — the value-add that consumes this), **#492** (GLM-DSA backend — owns the `evictGLMDsa` path's future), **#493** (network PP — separate KV/activation transport concern)
- Informs the FP8/INT8-KV decision raised in the parity matrix's *Quantized / low-precision KV cache* row

## Non-goals

- **Base-item parity only.** This proves the value-add survives a base-item layout change; it does not build the L3 value-add stack (#504-532) — those ride on top once exact-span-on-paged is proven.
- **Not building the paged allocator here** — this is the design+proof gate that unblocks it; the allocator is the sibling feat issue.
- **Track B scope: not chasing raw single-GPU throughput parity vs vLLM.** The point is that kernel-owned paged KV is *strictly better* via exact-span, not faster than PagedAttention at raw decode.
- Not implementing quantized KV — only deciding whether this design preserves or defers it.
- Not the cross-node KV transport / serializer (separate Track-B data-plane issue).

---
_Migrated from an internal tracker._

**Comment by anthony-chaudhary** [comment]
## Closing as **done** — GO verdict proven and shipped

All six acceptance criteria are met by committed code with a passing bit-exact proof. Evidence below; nothing in scope remains open (the paged allocator and L3 referee are explicit non-goals here — they ride on this gate).

### Acceptance criteria — all met
- [x] **Written design committed under `docs/`** — `docs/notes/EVICT-ON-PAGED-KV-DESIGN-2026-06-28.md` (commits c36cd6a9, 48068f07). Establishes the `== never-saw-it` guarantee survives paging because the single-rotation re-RoPE depends only on the survivor's `Kraw` and its **new logical index**, not on physical contiguity.
- [x] **Survivor re-RoPE-across-block-boundaries + COW-shared-block + `Kraw`-per-block** — design §3/§4. `Kraw` is a third per-block plane (in lock-step with K/V); there is *no* special "across a block boundary" case for the re-RoPE — only the re-index bookkeeping crosses boundaries. A survivor re-RoPE triggers a copy-on-write split of any shared block before mutating it.
- [x] **Prototype bit-identical to contiguous `Evict`** — `internal/model/paged_evict.go` + `paged_evict_test.go` (commit f007995b, kernel-audited *diff-witnessed*). `TestPagedEvictBitIdenticalToContiguous` evicts a non-block-aligned span `[2,5)` on a **deliberately non-contiguous** page table (the test fails the setup if the table is contiguous) under **layer-specific RoPE theta** (which would expose any composed-rotation drift), and asserts K, `Kraw`, V **and** next-token logits are byte-for-byte identical to `KVCache.Evict`. `max|Δ| = 0`.
- [x] **GLM-DSA path covered** — design §7; `internal/model/paged_glmdsa.go` pages each GLM-DSA row family (attention K/Kraw/V + learned-indexer K/Kraw) under its own block table. `TestPagedGLMDsaEvictBitIdenticalToContiguous` proves a mid-span evict + next decode step is bit-identical to the contiguous `evictGLMDsa`.
- [x] **FP8/INT8-KV interaction stated** — design §6: keep `Kraw` in **f32** as the invariant carrier (V may be quantized freely; K may be quantized for the read with a parallel f32 `Kraw` shadow). Quantized-K-**without** an f32 shadow cannot offer bit-exact eviction and must degrade to whole-prefix flush honestly. Quantized-K-with-exact-span is **deferred** with that written reason.
- [x] **Explicit go/no-go** — design §9: **GO.** Exact-span eviction survives paging at no new asymptotic cost (an O(survivors-after-span) intra-block compaction replaces the contiguous splice; the re-RoPE cost is unchanged). This is fak's structural differentiator: vLLM `reset_prefix_cache` / SGLang `flush_cache` pay zero for middle-span eviction only because they cannot do it (`enginecache.go` hardcodes `SupportsExactSpan=false`).

### Proof run (local, `go test ./internal/model`)
```
--- PASS: TestPagedEvictBitIdenticalToContiguous
--- PASS: TestPagedEvictCOWLeavesForkedParentUnchanged
--- PASS: TestPagedEvictGovernedAttestationCarriesRefereeDescriptor
--- PASS: TestPagedEvictGovernedRefusesUngovernedSpanBeforeMutation
--- PASS: TestPagedGLMDsaEvictBitIdenticalToContiguous
```

### Binding commits
- f007995b `feat(model): add bit-exact middle-span Evict on paged/block KV via a Kraw plane, plus paged reserve (#33)`
- 860d7d43 `feat(model): add governed paged-KV eviction with referee attestation (#33)`
- c36cd6a9 / 48068f07 — the design notes

### Unblocks
- **#34** (native paged allocator carrying `Evict`): GO — build the allocator with the `Kraw` plane from the start and carry `Evict` via the proven re-index + single-rotation-from-`Kraw` + COW pattern. (#34 still owes the paged-attention gather and serve-path wiring; this gate proved the eviction *math*.)
- **#27** (L3 KV-governance referee): native exact-span eviction is achievable on the paged tier and degrades honestly to whole-prefix flush on the ridden-engine path.
